File: GcdOfStrings.cpp
#include <iostream>
#include <string>

class Solution{
    public:
        std::string gcdOfString(std::string str1, std::string str2){
            return (str1+str2 == str2+str1) ? str1.substr(0, gcd(str1.length(), str2.length())) : "";
        }
    private:
        int gcd(int l1, int l2){
            return l2 == 0 ? l1 : gcd(l2, l1%l2);
        }
};

int main(){
    Solution s;
    std::cout << s.gcdOfString("ABCABC", "ABC") << std::endl;
    std::cout << s.gcdOfString("LEET", "CODE") <<std::endl;
}

File: LC 75/GCDOfStrings.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(gcdOfStrings("ABABAB", "AB"))
	fmt.Println(gcdOfStrings("LEET", "CODE"))
	fmt.Println(gcdOfStrings("ABCABC", "ABC"))
}

func gcdOfStrings(str1, str2 string) string {
	if str1+str2 != str2+str1 {
		return ""
	}
	var gcd func(len1, len2 int) int
	gcd = func(len1, len2 int) int {
		if len2 == 0 {
			return len1
		}
		return gcd(len2, len1%len2)
	}
	return str1[:gcd(len(str1), len(str2))]
}
